pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed

pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12773214?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12773214?buildTab=artifacts#/) on master @ [64dacab2808c5420e3fd99d2562e8ecb97b7742b](https://github.com/cockroachdb/cockroach/commits/64dacab2808c5420e3fd99d2562e8ecb97b7742b):

Fatal error:

```
panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING)) [recovered]
	panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING)) [recovered]
	panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING)) [recovered]
	panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING)) [recovered]
	panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING))

```
Stack: 

```
goroutine 15 [running]:
testing.tRunner.func1.2({0x8eb0640, 0xc002072990})
	GOROOT/src/testing/testing.go:1545 +0x366
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x630
panic({0x8eb0640?, 0xc002072990?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:133 +0x3ad
panic({0x8eb0640?, 0xc002072990?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0075d4788?, {0xcc81018, 0x11ddbf60})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x74
panic({0x8eb0640?, 0xc002072990?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0031b1e00, {0xcc81018, 0x11ddbf60})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:521 +0x25d
panic({0x8eb0640?, 0xc002072990?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MemoizeComparisonExprOp(0xc003ea4780)
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:492 +0x52a
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.NewTypedComparisonExpr({0xc0010411e8?, 0x40?}, {0xccb2f10?, 0xc00747c960}, {0xccb3450?, 0xc0020728d0})
	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:401 +0x1dd
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeCompareOp(0xc007a44240, 0x1077a440, {0xc00264a680, 0x2e, 0x34})
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:321 +0x2b8
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.init.scalarNoContext.func17(0xc0028d71a0?, {0xc0075d4d01?, 0x40?}, 0xc000e05500?, {0xc00264a680, 0x2e, 0x34})
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:68 +0x77
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeScalarSample(0x48ed89?, 0xc007a44240, {0x713ba5?, 0xe0?}, 0x1077a440, {0xc00264a680, 0x2e, 0x34}, 0x1)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:113 +0x29d
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeBoolExprContext(...)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:97
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeBoolExpr(...)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:84
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeWhere(0xc007a44240, {0xc00264a680, 0x2e, 0x34}, 0x1)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:1543 +0x17b
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelectClause(0xc007a44240, {0x0?, 0x0, 0x0}, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:671 +0xae5
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelect(0x48ed89?, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0})
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:832 +0xc5
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeSelect(0xc007a44240)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:784 +0x4e
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeStmt(0xc007a44240)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:26 +0x52
github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).Generate(0xc00305b6c0?)
	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/sqlsmith.go:184 +0x3b
github.com/cockroachdb/cockroach/pkg/sql/tests.GenerateAndCheckRedactedExplainsForPII(0xc00305b6c0, 0xc0018b2b70?, 0xa, 0xc0033a4140, 0xc003918a80)
	github.com/cockroachdb/cockroach/pkg/sql/tests/explain_test_util.go:38 +0x16d
pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL(0xc00305b6c0)
	pkg/ccl/testccl/sqlccl/sqlccl_test/pkg/ccl/testccl/sqlccl/explain_test.go:125 +0xebe
testing.tRunner(0xc00305b6c0, 0x99826b0)
	GOROOT/src/testing/testing.go:1595 +0x239
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x82b
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | [...repeated from below...]
* Wraps: (2) assertion failure
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.MemoizeComparisonExprOp
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:492
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.NewTypedComparisonExpr
*   | 	github.com/cockroachdb/cockroach/pkg/sql/sem/tree/expr.go:401
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeCompareOp
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:321
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.init.scalarNoContext.func17
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:68
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeScalarSample
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:113
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeBoolExprContext
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:97
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeBoolExpr
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/scalar.go:84
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeWhere
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:1543
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelectClause
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:671
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelect
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:832
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeSelect
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:784
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeStmt
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:26
*   | github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).Generate
*   | 	github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/sqlsmith.go:184
*   | github.com/cockroachdb/cockroach/pkg/sql/tests.GenerateAndCheckRedactedExplainsForPII
*   | 	github.com/cockroachdb/cockroach/pkg/sql/tests/explain_test_util.go:38
*   | pkg/ccl/testccl/sqlccl/sqlccl_test.TestExplainRedactDDL
*   | 	pkg/ccl/testccl/sqlccl/sqlccl_test/pkg/ccl/testccl/sqlccl/explain_test.go:125
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1595
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1650
* Wraps: (4) lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING))
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
    stopper.go:230: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /artifacts/tmp/_tmp/5c751b11a9f469cbdd173f9eebde1a6d/logTestExplainRedactDDL2937274931
    stopper.go:230: panic: lookup for ComparisonExpr (("t)ab5"."col3""_3")[refcursor] IS NOT DISTINCT FROM (e'pachycephalosaurushT\x01D9Nq')[string])[bool]'s CmpOp failed (IS NOT DISTINCT FROM(REFCURSOR,STRING))
--- FAIL: TestExplainRedactDDL (12.50s)
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #114705 pkg/ccl/testccl/sqlccl/sqlccl_test: TestExplainRedactDDL failed [C-test-failure O-robot T-sql-queries branch-release-23.2]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExplainRedactDDL.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33681

--- a/pkg/internal/sqlsmith/scalar.go
+++ b/pkg/internal/sqlsmith/scalar.go
@@ -204,7 +204,13 @@ func makeConstDatum(s *Smither, typ *types.T) tree.Datum {
 		if typ.Width() > 0 {
 			sv = util.TruncateString(sv, int(typ.Width()))
 		}
-		datum = tree.NewDString(sv)
+		if typ.Family() == types.RefCursorFamily {
+			// REFCURSOR is not compatible with the other string-like types, so make
+			// sure not to lose the type of the datum.
+			datum = tree.NewDRefCursor(sv)
+		} else {
+			datum = tree.NewDString(sv)
+		}
 	}
 	return datum
 }

